render/cross_section/cross_section_rendering_engine_4d.cpp

```diff
@@ -0,0 +1,133 @@
+		}
+
+		Projection modelview_basis = modelview_basises[mesh_index];
+		Vector4 modelview_origin = modelview_origins[mesh_index];
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_origin", modelview_origin);
+		// Can't pass a mat4 through instance uniforms, need to break up into columns.
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_x", modelview_basis.columns[0]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_y", modelview_basis.columns[1]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_z", modelview_basis.columns[2]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_w", modelview_basis.columns[3]);
+
+		instance_index++;
+	}
+	while (_instances_3d.size() > instance_index) {
+		RS::get_singleton()->free(_instances_3d.pop_back());
+	}
+}
+
+void CrossSectionRenderingEngine4D::update_camera() {
+	// TODO 3D preview is broken, mesh is not drawn relative to preview camera.
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	if (_cross_section_camera.is_null()) {
+		ERR_FAIL_NULL(get_viewport());
+		_cross_section_camera = RS::get_singleton()->camera_create();
+		RS::get_singleton()->viewport_attach_camera(get_viewport()->get_viewport_rid(), _cross_section_camera);
+	}
+	// Only setting final 3D->2D projection stuff here. Can't pass the full Transform4D or even the basis as the camera's transform because it expects a Transform3D.
+	ERR_FAIL_NULL(get_camera());
+	Camera4D *camera = get_camera();
+	// Projection type is a bitmask of 3D and 4D projection. Ignoring 4D projection because we're cross-sectioning instead.
+	if (camera->get_projection_type() == Camera4D::PROJECTION4D_ORTHOGRAPHIC) {
+		RS::get_singleton()->camera_set_orthogonal(_cross_section_camera, camera->get_orthographic_size(), camera->get_clip_near(), camera->get_clip_far());
+	} else if (bool(camera->get_projection_type() & Camera4D::PROJECTION4D_PERSPECTIVE_3D)) {
+		RS::get_singleton()->camera_set_perspective(_cross_section_camera, Math::rad_to_deg(camera->get_field_of_view_3d()), camera->get_clip_near(), camera->get_clip_far());
+	} else if (bool(camera->get_projection_type() & Camera4D::PROJECTION4D_PERSPECTIVE_4D)) {
+		WARN_PRINT_ONCE("4D Perspective settings are ignored when using the Cross-section renderer. Use PESPECTIVE_3D or ORTHOGRAPHIC.");
+	}
+}
+
+void CrossSectionRenderingEngine4D::setup_for_viewport() {
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	ERR_FAIL_NULL(get_viewport());
+	Ref<World3D> world = get_viewport()->find_world_3d();
+	// This will move instances between World3Ds, so adding a new viewport will remove all the instances from the previous viewport.
+	// TODO Per viewport instance cache?
+	if (world.is_valid()) {
+		RID scenario = world->get_scenario();
+		for (RID instance : _instances_3d) {
```

Comment:
The iterator of a `TypedArray` returns `Variant`. I'm a bit surprised this works in-engine. But anyway, this is internal data, so we should use `Vector` instead, which is type-safe and packed, in contrast to `TypedArray` which can be passed to GDScript and GDExtension (that's why the `RenderingEngine4D` API uses `TypedArray`, to allow implementing renderers outside the engine if needed).

Reply:
Yeah, makes sense. Switched to `vector<RID>`